src/Microsoft.Health.Fhir.SqlServer/Features/ChangeFeed/SqlServerFhirResourceChangeDataStore.cs

```diff
@@ -0,0 +1,159 @@
+                        sqlCommand.Parameters.AddWithValue("@startId", SqlDbType.BigInt).Value = startId;
+                        sqlCommand.Parameters.AddWithValue("@pageSize", SqlDbType.SmallInt).Value = pageSize;
+                        using (SqlDataReader sqlDataReader = await sqlCommand.ExecuteReaderAsync(CommandBehavior.SequentialAccess, cancellationToken))
+                        {
+                            while (await sqlDataReader.ReadAsync(cancellationToken))
+                            {
+                                listResourceChangeData.Add(new ResourceChangeData(
+                                    id: (long)sqlDataReader["Id"],
+                                    timestamp: DateTime.SpecifyKind((DateTime)sqlDataReader["Timestamp"], DateTimeKind.Utc),
+                                    resourceId: (string)sqlDataReader["ResourceId"],
+                                    resourceTypeId: (short)sqlDataReader["ResourceTypeId"],
+                                    resourceVersion: (int)sqlDataReader["ResourceVersion"],
+                                    resourceChangeTypeId: (byte)sqlDataReader["ResourceChangeTypeId"]));
+                            }
+                        }
+
+                        return listResourceChangeData;
+                    }
+                }
+            }
+            catch (SqlException ex)
+            {
+                switch (ex.Number)
+                {
+                    case TIMEOUTEXPIRED:
+                        throw new TimeoutException(ex.Message, ex);
+                    default:
+                        _logger.LogError(ex, string.Format(Resources.SqlExceptionOccurredWhenFetchingResourceChanges, ex.Number));
+                        throw;
+                }
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, Resources.ExceptionOccurredWhenFetchingResourceChanges);
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Get resource types as a dictionary.
+        /// </summary>
+        /// <param name="cancellationToken">Cancellation Token.</param>
+        /// <returns>Returns a resource type map.</returns>
+        /// <exception cref="System.InvalidOperationException">Thrown when a method call is invalid for the object's current state.</exception>
+        /// <exception cref="Microsoft.Data.SqlClient.SqlException">Thrown when SQL Server returns a warning or error.</exception>
+        /// <exception cref="System.TimeoutException">Thrown when the time allotted for a process or operation has expired.</exception>
+        /// <exception cref="System.Exception">Thrown when errors occur during execution.</exception>
+        public async Task<IDictionary<short, string>> GetResourceTypeMapAsync(CancellationToken cancellationToken)
+        {
+            var resourceTypeIdToNameMapping = new Dictionary<short, string>();
```

Comment:
>resourceTypeIdToNameMapping [](http://example.com/codeflow?start=16&length=27)

Cache it please.

That collection is constant for service life duration.

Reply:
Sure, resource types will be cached on a caller side in the publish event worker.

Reply:
What if they don't?  
Why rely on upstream when we can cache it here?

Reply:
I think that a GetResourceTypeMapAsync is a DAO function for the SQL server which gets resource types from the database. Caching should be on the caller's side.

Reply:
Yes, it's a DAO function and it has specific behaviour of returning same results again and again. Pushing responsibility to account to this specific behaviour on caller is not a good thing to do, because caller can ignore to do it, and just bombard us with requests.

It's better to incapsulate this logic inside method. 

As alternative we can always make this one time call function and throw exception on second call to force caller to call it only once. But that's not a good solution comparing to caching results inside function.

Reply:
That doesn't sound right to me Ivan. Let's say you have a Web App and SQL server. Would you cache from your middle tier or your DOA layer.

Reply:
There should be a difference between handling generic cases and specific cases, right?

If I know what collection is same all the time, I'll cache in DAO, because it's I know it's same all the time, and I'm not sure person who works in middle tier know about it.


[https://github.com/microsoft/fhir-server/blob/ecf39a2888cb8449b1085867288153f3f8a8b607/src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs#L356](https://github.com/microsoft/fhir-server/blob/ecf39a2888cb8449b1085867288153f3f8a8b607/src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs#L356)

We already cache this collection by reading it once.


Which also brings another issue, before first request to database that collection would be empty.

So if event agent is hosted service, it wouldn't be right to cache it.

Any chance you can event agent can use ISqlServerFhirModel to get dictionary? (I know there is no dictionary in interface, but we can extend interface) to protect from been called in wrong time?

Reply:
Added caching.